File: C/Algorithms/String/longestPalindrome.c
#include <string.h>
#include "isPalindrome.c"

 /**
 * Finds the longest palindrome in a given string
 *
 * @param char *string
 * @return char *
 */
char * findLongestPalindrome(char *string)
{
    if (isPalindromeIteration(string)) {
        return string;
    }

    int stringLength = strlen(string);
    char  *longestPalindrome = "";
    int longestPalindromeLength = 1;

    for (int start = 0;start < stringLength - 1;start++) {
        for (int end = start + 2;end < stringLength;end++) {
            char *subString = "";

            for (int i = start, j = 0;i <= end;i++, j++) {
                subString[j] = string[i];
            }

            if (isPalindromeIteration(subString)) {
                int tempSubStringLength = strlen(subString);
                if (tempSubStringLength > longestPalindromeLength) {
                    longestPalindrome = subString;
                    longestPalindromeLength = tempSubStringLength;
                }
            }
        }
    }

    if (longestPalindromeLength == 1) {
        char *result = "";
        result[0] = string[0];

        return result;
    }

    return longestPalindrome;
}

File: Java/Algorithms/String/LongestPalindrome.java
/**
 * Class LongestPalindrome
 */
public class LongestPalindrome {

    /**
     * Finds the longest palindrome in a given string
     *
     * @param str a given string in which the longest palindrome must be found
     * @return String
     */
    public String findLongestPalindrome(String str) {
        IsPalindrome isPalindrome = new IsPalindrome();

        if (isPalindrome.isPalindromeIteration(str)) {
            return str;
        }

        int stringLength = str.length();
        String longestPalindrome = "";
        int longestPalindromeLength = 1;

        for (int start = 0;start < stringLength - 1;start++) {
            for (int end = start + 3;end < stringLength + 1;end++) {
                String subString = str.substring(start, end);
                if (isPalindrome.isPalindromeIteration(subString)) {
                    int tempSubStringLength = subString.length();
                    if (tempSubStringLength > longestPalindromeLength) {
                        longestPalindrome = subString;
                        longestPalindromeLength = tempSubStringLength;
                    }
                }
            }
        }

        if (longestPalindromeLength == 1) {
            return String.valueOf(str.charAt(0));
        }

        return longestPalindrome;
    }
}
